clients/pkg/promtail/client/client_writeto.go

```diff
@@ -0,0 +1,89 @@
+package client
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/prometheus/prometheus/tsdb/record"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/ingester/wal"
+	"github.com/grafana/loki/pkg/util"
+)
+
+// clientWriteTo implements a wal.WriteTo that re-builds entries with the stored series, and the received entries. After,
+// sends each to the provided Client channel.
+type clientWriteTo struct {
+	series     map[chunks.HeadSeriesRef]model.LabelSet
+	seriesLock sync.RWMutex
+
+	// seriesSegment keeps track of the last segment in which the series pointed by each key in this map was seen. Keeping
+	// this in a separate map avoids unnecessary contention.
+	//
+	// Even though it doesn't present a difference right now according to benchmarks, it will help when we introduce other
+	// calls from the wal.Watcher to the wal.WriteTo like `UpdateSeriesSegment`.
+	seriesSegment     map[chunks.HeadSeriesRef]int
+	seriesSegmentLock sync.RWMutex
+
+	logger   log.Logger
+	toClient chan<- api.Entry
+}
+
+// newClientWriteTo creates a new clientWriteTo
+func newClientWriteTo(toClient chan<- api.Entry, logger log.Logger) *clientWriteTo {
+	return &clientWriteTo{
+		series:        make(map[chunks.HeadSeriesRef]model.LabelSet),
+		seriesSegment: make(map[chunks.HeadSeriesRef]int),
+		toClient:      toClient,
+		logger:        logger,
+	}
+}
+
+func (c *clientWriteTo) StoreSeries(series []record.RefSeries, segment int) {
+	c.seriesLock.Lock()
+	defer c.seriesLock.Unlock()
+	c.seriesSegmentLock.Lock()
+	defer c.seriesSegmentLock.Unlock()
+	for _, seriesRec := range series {
+		c.seriesSegment[seriesRec.Ref] = segment
+		// maybe we can avoid doing this below if already present
+		labels := util.MapToModelLabelSet(seriesRec.Labels.Map())
+		c.series[seriesRec.Ref] = labels
+	}
+}
+
+// SeriesReset will delete all cache entries that were last seen in segments numbered equal or lower than segmentNum
```

Comment:
I think this comment is currently incorrect, looks to me like we're just doing <

Reply:
Changing the if condition to `<=` since when `SeriesReset(n)` is called, it means segment n has been deleted, and because all segments keep an order in which they are written, every other before that one should have been reclaimed as well.